quickwit-search/src/collector.rs

```diff
@@ -64,11 +67,30 @@ impl SortingFieldComputer {
                     SortOrder::Asc => u64::MAX - field_val,
                 }
             }
-            SortingFieldComputer::SortByDocId => 0u64,
+            SortingFieldComputer::DocId => 0u64,
+            SortingFieldComputer::Score { order } => {
+                // Since we need a u64 value to return, we *transform* the BM25 score into u64.
+                // The transformation preserves the order.
+                // (https://lemire.me/blog/2020/12/14/converting-floating-point-numbers-to-integers-while-preserving-order/)
+                let u32_score = u32::from_le_bytes(score.to_le_bytes());
+                let u64_score = sign_flip(u32_score);
+                match order {
+                    SortOrder::Desc => u64_score,
+                    SortOrder::Asc => u64::MAX - u64_score,
+                }
+            }
         }
     }
 }
 
+/// Takes a u32 representation of a f32 and negates all bits if the most
+/// significant bit is set.
+fn sign_flip(value: u32) -> u64 {
+    let mut mask = (value as i64 >> 63) as u64;
```

Comment:
The conversion from `u32` to `i64` will screw the order up. There are two ways to go about this:
```rust
/// Converts a float to an unsigned integer while preserving order.
/// See https://lemire.me/blog/2020/12/14/converting-floating-point-numbers-to-integers-while-preserving-order/
fn f32_to_u64(value: f32) -> u64 {
    let value_u32 = u32::from_le_bytes(value.to_le_bytes());
    let mut mask = (value_u32 as i32 >> 31) as u32;
    mask |= 0x80000000;
    (value_u32 ^ mask) as u64
}
```
OR
```rust
fn f32_to_u64_bora(value: f32) -> u64 {
    let value_u64 = u64::from_le_bytes((value as f64).to_le_bytes());
    let mut mask = (value_u64 as i64 >> 63) as u64;
    mask |= 0x8000000000000000;
    value_u64 ^ mask
}
```
According to Godbolt, the former generates more efficient machine code, so let's go with that along with a bunch of unit tests such as:
```rust
assert!(f32_to_u64(-1.0); < f32_to_u64(-0.9999999));
```
This is a great use case for property testing.